src/python/pants/core/util_rules/environments.py

```diff
@@ -625,8 +626,38 @@ async def get_target_for_environment_name(
     return EnvironmentTarget(tgt)
 
 
+@rule_helper
+async def _maybe_add_docker_image_id(image_name: str, platform: Platform) -> str:
+    # If the image name appears to be just an image ID, just return it as-is.
+    if image_name.startswith("sha256:"):
+        return image_name
+
+    # If the image name contains an appended image ID, then use it.
+    if "@" in image_name:
+        image_name_part, _, maybe_image_id = image_name.rpartition("@")
+        if not maybe_image_id.startswith("sha256:"):
+            raise ValueError(
+                f"Docker image `{image_name}` contains an image ID component, but it does not appear to be a image SHA-256 hash."
+            )
+        return image_name
+
+    # Otherwise, resolve the image name to an image ID and append the applicable component to the image name.
+    resolve_result = await Get(
+        DockerResolveImageResult,
+        DockerResolveImageRequest(
+            image_name=image_name,
+            platform=platform.name,
+        ),
+    )
+
+    # TODO: Consider appending the correct image ID to the existing image name so error messages about the
```

Comment:
This needs further exploration. Merely returning `f"{image_name}@{resolve_result.image_id}"` failed to work in testing with the Docker command runner due to the overall image ID being different than the ID appended to tag names.

For example, with `python:3.9`, `docker image inspect python:3.9` returns an `Id` of `sha256:d732b9878117efd35e3ac57ec9db1daefb52e73424878b91c2bf006a44876b2e` and a `RepoDigests` of `python@sha256:cfcc9ef77b6cf87f57327aacc6f9b50c7cdb4d7dd93662a36549f05b7403cd47`. The latter method is probably the ID that needs to be used.

Reply:
Probably worth a tracking ticket

Reply:
Done.

Reply:
Hm, this feels like more like a blocking issue than a new ticket... if we use only shas then error messages will not be terribly useful.

It seems like it would be straightforward to use `python@sha256:cfcc9ef77b6cf87f57327aacc6f9b50c7cdb4d7dd93662a36549f05b7403cd47` in this change?